### DISCUSSION TITLE
status of `CAPNP_ALLOW_UNALIGNED`

### DISCUSSION THREAD

**Discussion by SamZhangQingChuan** [capnproto/capnproto#1878] [Q&A] [answered]
In this [issue](https://github.com/capnproto/capnproto/issues/1150#issuecomment-759571051), you said that `CAPNP_ALLOW_UNALIGNED` should be removed. I am wondering is it still planned to be removed?

Also, you said that a copy is inevitable if the original buffer is not aligned. I know it's UB to do unaligned access, but actually how unsafe is it? I am asking because it would require quite some work of refactoring and testing to add the functionality of an extra copy. Thanks!

**Comment by kentonv** [comment, marked as answer]
> I am wondering is it still planned to be removed?

I didn't have any immediate plans. I think it's unlikely to be removed in version 1, but likely in version 2.

> but actually how unsafe is it?

It's actually unsafe. For example:
- Although x86 _generally_ doesn't require alignment, it does feature some instructions that require alignment, especially SIMD instructions. The compiler can and does use these instructions sometimes when optimizing. It's hard to know when it will use them.
- The compiler can and occasionally does assume that the lower bits of a pointer are zero.

**Reply by SamZhangQingChuan** [reply]
Thanks!
